compiler/rustc_mir_build/src/build/expr/as_constant.rs

```diff
@@ -1,43 +1,153 @@
 //! See docs in build/expr/mod.rs
 
 use crate::build::Builder;
-use rustc_middle::mir::interpret::{ConstValue, Scalar};
+use crate::thir::constant::parse_float;
+use rustc_ast::ast;
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::interpret::{
+    Allocation, ConstValue, LitToConstError, LitToConstInput, Scalar,
+};
 use rustc_middle::mir::*;
 use rustc_middle::thir::*;
-use rustc_middle::ty::CanonicalUserTypeAnnotation;
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::{self, CanonicalUserTypeAnnotation, Ty, TyCtxt};
+use rustc_target::abi::Size;
 
 impl<'a, 'tcx> Builder<'a, 'tcx> {
     /// Compile `expr`, yielding a compile-time constant. Assumes that
     /// `expr` is a valid compile-time constant!
     crate fn as_constant(&mut self, expr: &Expr<'tcx>) -> Constant<'tcx> {
+        debug!("expr: {:#?}", expr);
+        // FIXME: Maybe we should try to evaluate here and only create an `Unevaluated`
+        // constant in case the evaluation fails. Need some evaluation function that
+        // allows normalization to fail.
```

Comment:
that comment seems incorrect, `const_eval_resolve` allows normalization to fail, it just causes an ICE when used in the incorrect `param_env`.